dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java

```diff
@@ -0,0 +1,182 @@
+        WorkerGroupTaskDispatcher looper =
+                dispatchWorkerMap.computeIfAbsent(workerGroup,
+                        k -> new WorkerGroupTaskDispatcher(workerGroup, taskExecutorClient));
+        looper.start();
+    }
+
+    /**
+     * Stop all workerGroupTaskDispatchWaitingQueueLooper
+     */
+    @Override
+    public void close() throws Exception {
+        log.info("WorkerGroupTaskDispatcherManager start close");
+        for (Map.Entry<String, WorkerGroupTaskDispatcher> entry : dispatchWorkerMap.entrySet()) {
+            try {
+                entry.getValue().close();
+            } catch (Exception e) {
+                log.error("stop worker group error", e);
+            }
+        }
+        log.info("WorkerGroupTaskDispatcherManager closed");
+    }
+
+    @Override
+    public void onWorkerGroupAdd(List<WorkerGroup> workerGroups) {
+        for (WorkerGroup workerGroup : workerGroups) {
+            this.addWorkerGroup(workerGroup.getName());
+        }
+    }
+
+    @Override
+    public void onWorkerGroupChange(List<WorkerGroup> workerGroups) {
+        String workerGroupsString = workerGroups.stream()
+                .map(WorkerGroup::getName)
+                .collect(Collectors.joining(", "));
+        log.info("Worker groups: {}", workerGroupsString);
+    }
+
+    @Override
+    public void onWorkerGroupDelete(List<WorkerGroup> workerGroups) {
+        for (WorkerGroup workerGroup : workerGroups) {
+            try {
+                this.deleteWorkerGroup(workerGroup.getName());
+            } catch (Exception e) {
+                log.error("stop worker group error", e);
+            }
+        }
+    }
+
+    private void checkDeleteDispatchWorkerComplete() {
+        for (Map.Entry<String, WorkerGroupTaskDispatcher> entry : dispatchWorkerMap.entrySet()) {
+            String workerGroup = entry.getKey();
+            WorkerGroupTaskDispatcher workerGroupTaskDispatcher = entry.getValue();
+            switch (workerGroupTaskDispatcher.getStatus()) {
+                case CLOSING:
+                    try (WorkerGroupTaskDispatcher ignored = workerGroupTaskDispatcher) {
+                        log.info("try to delete worker group {}", workerGroup);
+                    } catch (Exception e) {
+                        log.error("stop worker group error", e);
+                    }
+                    break;
+                case CLOSED:
+                    try (WorkerGroupTaskDispatcher ignored = dispatchWorkerMap.remove(workerGroup)) {
+                        log.info("success remove worker group {}", workerGroup);
+                    } catch (Exception e) {
+                        log.error("stop worker group error", e);
+                    }
+                    break;
+                default:
+                    log.debug("worker group {} status {}", workerGroup, workerGroupTaskDispatcher.getStatus());
+                    break;
+            }
+        }
+    }
```

Comment:
Once a dispatcher's status is closed, then it might still exist task in its queue, I don't see any control about this.